### PR TITLE
Implement readlink command

### PR DESCRIPTION
**Pull request by bkochendorfer**
## Description

`readlink` command with flags '-f' and '-v'. Also update the roadmap
to reflect the addition of this command and what flags are left to
implement.

## Testing
Added tests covering the usage and combination of different flags as implemented. Verified similiar behavior, for implemented flags, to `readlink (GNU coreutils) 8.25`. I cribbed the test style from an existing command.

### CODE REVIEW

File: cmds/readlink/readlink.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// readlink display value of symbolic link file
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: super nit: end sentences with a period.

File: cmds/readlink/readlink.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"path/filepath"
+)
+
+var (
+	follow  = flag.Bool("f", false, "follow recursively")
+	verbose = flag.Bool("v", false, "report error messages")
+)
+
+func isLink(file string) error {
+	var err error
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: no need to declare err here, the line below already derives its type.

File: cmds/readlink/readlink.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	f, err := os.Lstat(file)
+
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: super nit: keep the if-statement close to the Lstat.

```go
f, err := os.Lstat(file)
if err != nil {
  return err
}
```

File: cmds/readlink/readlink.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		return err
+	}
+
+	// If it is not a symbolic link return an error
+	if f.Mode()&os.ModeSymlink == 0 {
+		return fmt.Errorf("%s Invalid argument", file)
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: I think a more descriptive error message would be nice here, and using %q for quotes around the path:

```go
if m := f.Mode(); m&os.ModeSymlink == 0 {
  return fmt.Errorf("%q is not a symlink, but a %s", file, m)
}
```

File: cmds/readlink/readlink_test.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+func TestReadlink(t *testing.T) {
+
+	// Create an empty directory
+	tmpDir, execPath := testutil.CompileInTempDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	// Creating here to utilize path in tests
+	testDir := filepath.Join(tmpDir, "readLinkDir")
+	os.Mkdir(testDir, 0700)
```
Comment (hugelgupf) [inline cmds/readlink/readlink_test.go]: check error? here and below

File: cmds/readlink/readlink_test.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		},
+		{
+			flags:      []string{"-v", "multilinks", "f1symlink", "f2"},
+			out:        fmt.Sprintf("%s/%sf1\n", testDir, "f1symlink\n"),
+			stdErr:     "readlink: f2 Invalid argument\n",
+			exitStatus: 1,
+		},
+	}
+	// Createfiles.
+	os.Create("f1")
```
Comment (hugelgupf) [inline cmds/readlink/readlink_test.go]: ditto, check error, etc and below

File: cmds/readlink/readlink.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	return err
+}
+
+func readFile(file string) error {
+	var path string
+	var err error
+
+	err = isLink(file)
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: super nit: it's customary to embed this in the if-statement, especially if nothing else is declared:

```go
if err := isLink(file); err != nil {
  return err
}
```

File: cmds/readlink/readlink.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	return err
+}
+
+func main() {
+	flag.Parse()
+
+	exitStatus := 0
+
+	for _, file := range flag.Args() {
+		err := readFile(file)
```
Comment (hugelgupf) [inline cmds/readlink/readlink.go]: similarly, you can embed this in the if, and get rid of the empty line:

```go
if err := readFile(file); err != nil {
  if *verbose {
    fmt.Fprintf(os.Stderr, "readlink: %v\n", err)
  }
  exitStatus = 1
}
```

(notice how I changed %s to %v and dropped .Error() -- it's functionally equivalent, but shorter)

File: cmds/readlink/readlink_test.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+)
+
+type Test struct {
+	flags      []string
+	out        string
+	stdErr     string
+	exitStatus int
+}
+
+func TestReadlink(t *testing.T) {
```
Comment (hugelgupf) [inline cmds/readlink/readlink_test.go]: Awesome tests!!

File: cmds/readlink/readlink_test.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	"github.com/u-root/u-root/pkg/testutil"
+)
+
+type Test struct {
```
Comment (hugelgupf) [inline cmds/readlink/readlink_test.go]: nit: you need not export `Test`, this could just be `test`.

File: cmds/readlink/readlink.go (hunk 2 above)
Comment (rminnich) [reply cmds/readlink/readlink.go]: Actually, the easiest way to see if a link is a link is readlink and look for an error.
So just do the readlink below, and see if the error is "this is not a symlink". If it is not, then
return that error. If *follow is false, you're done -- you got the value in Readlink.

doing a stat and a readlink is actually racy, and believe it or not, I've had it happen.

**Comment by bkochendorfer** [comment]
@rminnich @hugelgupf Thank you for the great feedback! Updated based on your recommendations.

**Comment by bkochendorfer** [comment]
@rminnich appreciate the feedback. Added the `init` and `var exitStatus int` variable init.

**Comment by rminnich** [comment]
Go for it. This is done!

now please review my DHCP PR :-)

**Pull request merged**